vmm/src/signal_handler.rs

```diff
@@ -22,53 +21,63 @@ const SI_OFF_SYSCALL: isize = 6;
 
 const SYS_SECCOMP_CODE: i32 = 1;
 
-// This no longer relies on sys_util::register_signal_handler(), which is a lot weirder than it
-// should be (at least for this use case). Also, we want to change the sa_mask field of the
-// sigaction struct.
-/// Sets up the specified signal handler for `SIGSYS`.
-pub fn setup_sigsys_handler() -> Result<(), io::Error> {
-    // Safe, because this is a POD struct.
-    let mut sigact: sigaction = unsafe { mem::zeroed() };
-    sigact.sa_flags = libc::SA_SIGINFO;
-    sigact.sa_sigaction = sigsys_handler as usize;
-
-    // We set all the bits of sa_mask, so all signals are blocked on the current thread while the
-    // SIGSYS handler is executing. Safe because the parameter is valid and we check the return
-    // value.
-    if unsafe { sigfillset(&mut sigact.sa_mask as *mut sigset_t) } < 0 {
-        return Err(io::Error::last_os_error());
+/// Signal handler for `SIGSYS`.
+///
+/// Increments the `seccomp.num_faults` metric, logs an error message and terminates the process
+/// with a specific exit code.
+///
+extern "C" fn sigsys_handler(num: c_int, info: *mut siginfo_t, _unused: *mut c_void) {
+    // Safe because we're just reading some fields from a supposedly valid argument.
+    let si_signo = unsafe { (*info).si_signo };
+    let si_code = unsafe { (*info).si_code };
+
+    // Sanity check. The condition should never be true.
+    if num != si_signo || num != SIGSYS || si_code != SYS_SECCOMP_CODE as i32 {
+        // Safe because we're terminating the process anyway.
+        unsafe { _exit(i32::from(super::FC_EXIT_CODE_UNEXPECTED_ERROR)) };
     }
 
-    // Safe because the parameters are valid and we check the return value.
-    if unsafe { sigaction(libc::SIGSYS, &sigact, null_mut()) } < 0 {
-        return Err(io::Error::last_os_error());
+    // Other signals which might do async unsafe things incompatible with the rest of this
+    // function are blocked due to the sa_mask used when registering the signal handler.
+    let syscall = unsafe { *(info as *const i32).offset(SI_OFF_SYSCALL) as usize };
+    METRICS.seccomp.num_faults.inc();
+    error!(
```

Comment:
Can we log stuff in signal handler?

Reply:
I think so. [`write` is safe](http://man7.org/linux/man-pages/man7/signal-safety.7.html), and the formatting [doesn't use any of the unsafe `sprintf` family](https://github.com/rust-lang/rust/blob/master/src/libcore/fmt/mod.rs).